src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -149,6 +141,206 @@ public interface IFileWatchingToken
+
+            public static WatcherOperation UnwatchDirectories(List<uint> cookies)
+                => new(Kind.UnwatchDirectories, cookies);
+
+            public static WatcherOperation UnwatchFiles(IEnumerable<IFileWatchingToken> tokens)
+                => new(Kind.UnwatchFiles, tokens);
+
+            public static WatcherOperation UnwatchFile(IFileWatchingToken token)
+                => new(Kind.UnwatchFile, token);
+
+            public bool TryCombineWith(in WatcherOperation other, out WatcherOperation combined)
+            {
+                if (other._kind == Kind.None)
+                {
+                    combined = this;
+                    return true;
+                }
+                else if (_kind == Kind.None)
+                {
+                    combined = other;
+                    return true;
+                }
+
+                combined = default;
+                return false;
+            }
+
+            public async ValueTask ApplyAsync(IVsAsyncFileChangeEx service, CancellationToken cancellationToken)
+            {
+                switch (_kind)
+                {
+                    case Kind.None:
+                        return;
+
+                    case Kind.WatchDirectory:
+                        var cookie = await service.AdviseDirChangeAsync(_directory, watchSubdirectories: true, _sink, cancellationToken).ConfigureAwait(false);
+                        _cookies.Add(cookie);
+
+                        if (_filter != null)
+                            await service.FilterDirectoryChangesAsync(cookie, new string[] { _filter }, cancellationToken).ConfigureAwait(false);
+
+                        return;
+
+                    case Kind.WatchFile:
+                        ((Context.FileWatchingToken)_token).Cookie = await service.AdviseFileChangeAsync(_directory, _fileChangeFlags, _sink, cancellationToken).ConfigureAwait(false);
+                        return;
+
+                    case Kind.UnwatchFile:
+                        await service.UnadviseFileChangeAsync(((Context.FileWatchingToken)_token).Cookie!.Value, cancellationToken).ConfigureAwait(false);
+                        return;
+
+                    case Kind.UnwatchDirectories:
+                        Contract.ThrowIfFalse(_cookies is not null);
+                        foreach (var unwatchCookie in _cookies)
+                        {
+                            await service.UnadviseDirChangeAsync(unwatchCookie, cancellationToken).ConfigureAwait(false);
+                        }
+
+                        return;
+
+                    case Kind.UnwatchFiles:
+                        Contract.ThrowIfFalse(_tokens is not null);
+                        foreach (var token in _tokens)
+                        {
+                            await service.UnadviseFileChangeAsync(((Context.FileWatchingToken)token).Cookie!.Value, cancellationToken).ConfigureAwait(false);
+                        }
+
+                        return;
```

Comment:
i didn't actually validate this switch.

Reply:
ok, looking below, it seems like you just moved all that logic from down below into here.  Again, i didn't review it, but i imagine it's was all a very straightforward move with mechanical transformations.